src/vulkan-renderer/wrapper/vma.cpp

```diff
@@ -0,0 +1,78 @@
+#include "inexor/vulkan-renderer/wrapper/vma.hpp"
+
+#define VMA_IMPLEMENTATION
+
+// It makes memory of all new allocations initialized to bit pattern 0xDCDCDCDC.
+// Before an allocation is destroyed, its memory is filled with bit pattern 0xEFEFEFEF.
+// Memory is automatically mapped and unmapped if necessary.
+#define VMA_DEBUG_INITIALIZE_ALLOCATIONS 1
+
+// Enforce specified number of bytes as a margin before and after every allocation.
+#define VMA_DEBUG_MARGIN 16
+
+// Enable validation of contents of the margins.
+#define VMA_DEBUG_DETECT_CORRUPTION 1
+
+#include <vma/vk_mem_alloc.h>
+
+namespace inexor::vulkan_renderer::wrapper {
+
+VulkanMemoryAllocator::VulkanMemoryAllocator(VulkanMemoryAllocator &&other) noexcept : vma_allocator(std::exchange(other.vma_allocator, nullptr)) {}
+
+VulkanMemoryAllocator::VulkanMemoryAllocator(const VkInstance instance, const VkDevice device, const VkPhysicalDevice graphics_card) {
+    assert(device);
+
+    spdlog::debug("Initialising Vulkan memory allocator.");
+
+    // Make sure to set the root directory of this repository as working directory in the debugger!
+    // Otherwise, VMA won't be able to open this allocation replay file for writing.
+    const std::string vma_replay_file = "vma-replays/vma_replay.csv";
+
+    spdlog::debug("Opening VMA memory recording file for writing.");
+
+    std::ofstream replay_file_test;
+    replay_file_test.open(vma_replay_file, std::ios::out);
+
+    // Check if we can open the csv file.
+    // This causes problems when the debugging path is set incorrectly!
+    if (!replay_file_test.is_open()) {
+        throw std::runtime_error("Could not open VMA replay file " + vma_replay_file + " !");
+    }
+
+    spdlog::debug("VMA memory recording file opened successfully.");
+
+    replay_file_test.close();
```

Comment:
This shouldn't be needed.

Reply:
It is better to close as soon as possible.

Reply:
But it is better to close as soon as possible and personally i don't like the scope method.

Reply:
I would like to be explicit with this as well.